src/operator/join.rs

```diff
@@ -793,107 +653,217 @@ mod test {
                 let paths_inverted_indexed: Stream<_, OrdIndexedZSet<usize, usize, isize>> = paths_inverted.index();
                 let edges_indexed: Stream<_, OrdIndexedZSet<usize, usize, isize>> = edges.index();
 
-                let paths = edges.plus(&paths_inverted_indexed.join_incremental_nested(&edges_indexed, |_via, from, to| (*from, *to)))
-                    .distinct_incremental_nested();
+                let paths = edges.plus(&paths_inverted_indexed.join_trace::<NestedTimestamp32, _, _, _>(&edges_indexed, |_via, from, to| (*from, *to)))
+                    .distinct_trace();
                 paths_delayed.connect(&paths);
-                let output = paths.integrate();
-                Ok((
-                    vec![
-                        paths.condition(HasZero::is_zero),
-                        paths.integrate_nested().condition(HasZero::is_zero)
-                    ],
-                    output.export(),
-                ))
+                let output = paths.integrate_trace();
+                Ok(output.export())
             })
             .unwrap();
 
-            paths.integrate().distinct().inspect(move |ps| {
+            paths.consolidate::<OrdZSet<_, _>>().integrate().distinct().inspect(move |ps| {
                 assert_eq!(*ps, outputs.next().unwrap());
             })
         })
         .unwrap();
 
         for _ in 0..8 {
+            //eprintln!("{}", i);
             root.step().unwrap();
         }
     }
-    */
+
+    #[derive(Clone, Debug, Default, Ord, PartialOrd, Eq, PartialEq, DeepSizeOf)]
+    struct Label(pub usize, pub u16);
+
+    impl Display for Label {
+        fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+            write!(f, "L({},{})", self.0, self.1)
+        }
+    }
+
+    #[derive(Clone, Debug, Default, Ord, PartialOrd, Eq, PartialEq, DeepSizeOf)]
+    struct Edge(pub usize, pub usize);
+
+    impl Display for Edge {
```

Comment:
Maybe these should be library functions, I expect we'll have more graph tests.

Reply:
Yes, we'll have a graph library, but this implementation of `propagate` isn't actually correct because it doesn't apply `distinct` on each iteration.  This is done deliberately for testing purposes, so it won't be reusable as a library function.  As for the types, they'd have to be generic to be reusable.